File: 617.merge-two-binary-trees.py
#
# @lc app=leetcode id=617 lang=python3
#
# [617] Merge Two Binary Trees
#

# @lc code=start
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def mergeTrees(self, root1: TreeNode, root2: TreeNode) -> TreeNode:
        if not root1 and not root2:
            return None
        if not root1:
            return TreeNode(root2.val, root2.left, root2.right)
        if not root2:
            return TreeNode(root1.val, root1.left, root1.right)
        return TreeNode(root1.val+root2.val, self.mergeTrees(root1.left, root2.left), self.mergeTrees(root1.right, root2.right))
# @lc code=end

File: 617-merge-two-binary-trees/617-merge-two-binary-trees.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void helper(TreeNode* root1, TreeNode* root2){
        if(root1 && root2){
            root1->val += root2->val;
            if(root2->left && !root1->left){
                root1->left = root2->left;
            } else{
                helper(root1->left, root2->left);
            }
            if(root2->right && !root1->right){
                root1->right = root2->right;
            } else{
                helper(root1->right, root2->right);
            }
        } 
    }
    TreeNode* mergeTrees(TreeNode* root1, TreeNode* root2) {
        if(root1 && !root2){
            return root1;
        }
        if(root2 && !root1){
            return root2;
        }
        helper(root1, root2);
        return root1;
    }
};